### PR TITLE
chore: span origin

### PR DESCRIPTION
**Pull request by P403n1x87**
## Checklist

- [ ] Change(s) are motivated and described in the PR description
- [ ] Testing strategy is described if automated tests are not included in the PR
- [ ] Risks are described (performance impact, potential for breakage, maintainability)
- [ ] Change is maintainable (easy to change, telemetry, documentation)
- [ ] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [ ] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [ ] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [ ] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [ ] Title is accurate
- [ ] All changes are related to the pull request's stated goal
- [ ] Description motivates each change
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [ ] Testing strategy adequately addresses listed risks
- [ ] Change is maintainable (easy to change, telemetry, documentation)
- [ ] Release note makes sense to a user of the library
- [ ] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [ ] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/ext/__init__.py [hunk 1]
```diff
@@ -22,3 +22,17 @@ class SpanKind(object):
     SERVER = "server"
     PRODUCER = "producer"
     CONSUMER = "consumer"
+
+
+EXIT_SPAN_TYPES = frozenset(
```
Comment (shatzi) [inline ddtrace/ext/__init__.py:27]: is that a complete list? what is the plan on how to maintain this list?

File: ddtrace/debugging/_origin/span.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    def on_span_start(self, span: Span) -> None:
+        if span.span_type not in EXIT_SPAN_TYPES:
+            return
+
+        # Add call stack information to the exit span. Report only the part of
+        # the stack that belongs to user code.
+        # TODO: Add a limit to the number of frames to capture
+        seq = count(1)
+        for frame in frame_stack(sys._getframe(1)):
```
Comment (shatzi) [inline ddtrace/debugging/_origin/span.py:194]: did we decided to add the entire stack for exit spans?
if so, then how do we deal the frame numbering? currently is running numbers of user-code frames. but if we also report the entire stack, maybe those should match the frame depth?

File: ddtrace/debugging/_origin/span.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            take_snapshot=True,
+            limits=DEFAULT_CAPTURE_LIMITS,
+            condition=None,
+            condition_error_rate=0.0,
+            rate=float("inf"),
+        )
+
+
+@attr.s
+class ExitSpanProbe(LogLineProbe):
```
Comment (shatzi) [inline ddtrace/debugging/_origin/span.py:68]: do we want to keep exit probe as line probes? or do we want to move all exit span probes to be method probes?

**Review by shatzi — COMMENTED** [review]
looks good

File: ddtrace/bootstrap/preload.py [hunk 4]
```diff
@@ -66,11 +66,17 @@ def register_post_preload(func: t.Callable) -> None:
 
     symbol_db.bootstrap()
 
-if di_config.enabled or ed_config.enabled:
+# TODO: Do not enable RC features if not required
+if di_config.enabled or ed_config.enabled or config._trace_span_origin_enabled:
     from ddtrace.debugging import DynamicInstrumentation
 
     DynamicInstrumentation.enable()
```
Comment (shatzi) [inline ddtrace/bootstrap/preload.py:73]: enable DynamicInstrumentation.enable() without di_config.enabled still get probes?
I wonder if we need to split product features (get RC probes) from the library capabilities features (instrument methods/lines)

File: ddtrace/debugging/_origin/span.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+            module=module,
+            probe=t.cast(EntrySpanProbe, EntrySpanProbe.build(name=name, module=module, function=name)),
+        )
+
+    def __enter__(self):
+        super().__enter__()
+
+        root = ddtrace.tracer.current_root_span()
+        location = self.location
+        if root is None or root.get_tag("_dd.entry_location.file") is not None:
```
Comment (shatzi) [inline ddtrace/debugging/_origin/span.py:136]: java impl does something different, it add the tags to both the current active span and root span. not sure if we want to align that logic or not.

File: ddtrace/debugging/_origin/span.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+            return
+
+        # Add tags to the local root
+        root.set_tag_str("_dd.entry_location.file", location.file)
+        root.set_tag_str("_dd.entry_location.start_line", str(location.start_line))
+        root.set_tag_str("_dd.entry_location.end_line", str(location.end_line))
+        root.set_tag_str("_dd.entry_location.type", location.module)
+        root.set_tag_str("_dd.entry_location.method", location.name)
+
+        if config._trace_span_origin_enriched:
```
Comment (shatzi) [inline ddtrace/debugging/_origin/span.py:146]: why we take snapshot at entry and not on exit? I think it better to snapshot exit with locals and return values?
do we capture locals at the moment or just return here?

Comment (shatzi) [reply ddtrace/debugging/_origin/span.py:146]: second, I think we also need to look at _dd.p.debug flag for the enriched snapshots.

File: ddtrace/internal/telemetry/writer.py [hunk 7]
```diff
@@ -386,6 +386,12 @@ def _telemetry_entry(self, cfg_name: str) -> Tuple[str, str, _ConfigSource]:
         elif cfg_name == "_trace_sampling_rules":
             name = "trace_sampling_rules"
             value = str(item.value())
+        elif cfg_name == "_trace_span_origin_enabled":
+            name = "trace_span_origin_enabled"
+            value = "true" if item.value() else "false"
+        elif cfg_name == "_trace_span_origin_enriched":
+            name = "trace_span_origin_enriched"
```
Comment (shatzi) [inline ddtrace/internal/telemetry/writer.py:393]: maybe this config should be called the live_debugger? as what it should be in the future is to trigger snapshots based on _dd.p.debug flag

File: tests/debugging/origin/test_span.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        # Check for the expected tags on the entry span
+        assert entry.get_tag("_dd.entry_location.file") == str(Path(__file__).resolve())
+        assert entry.get_tag("_dd.entry_location.start_line") == str(min(lines))
+        assert entry.get_tag("_dd.entry_location.end_line") == str(max(lines))
+        assert entry.get_tag("_dd.entry_location.type") == __name__
+        assert entry.get_tag("_dd.entry_location.method") == "SpanProbeTestCase.test_span_origin.<locals>.entry_call"
+
+        # Check that we don't have span location tags on the middle span
+        assert middle.get_tag("_dd.entry_location.file") is None
+        assert middle.get_tag("_dd.exit_location.1.file") is None
```
Comment (shatzi) [inline tests/debugging/origin/test_span.py:52]: java have those indexed from - 0. we should sync on implementation details.

**Review by shatzi — APPROVED** [review]

File: ddtrace/debugging/_origin/span.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+        # the stack that belongs to user code.
+        seq = count(0)
+        for frame in frame_stack(sys._getframe(1), co_config.max_frame_depth):
+            code = frame.f_code
+            frame_origin = Path(code.co_filename)
+
+            if is_user_code(frame_origin):
+                print(frame_origin, "is user code")
+                n = next(seq)
+                span.set_tag_str(f"_dd.ld.code_origin.frames.{n}.file", str(frame_origin.resolve()))
```
Comment (shatzi) [inline ddtrace/debugging/_origin/span.py:215]: do we have method/type information here as well?

**Review by shatzi — APPROVED** [review]

File: ddtrace/debugging/_origin/span.py (hunk 9 above)
Comment (P403n1x87) [reply ddtrace/debugging/_origin/span.py:215]: no, unless we query the garbage collector, which is expensive :(

**Pull request closed without merging**